example3/OpenTelemetry.Metric.Api/Meter.cs

```diff
@@ -0,0 +1,173 @@
+
+        Gauge CreateGauge(string name);
+        Gauge CreateGauge(string name, Dictionary<string,string> labels);
+        Gauge CreateGauge(string name, params string[] dimNames);
+        Gauge CreateGauge(string name, Dictionary<string,string> labels, params string[] dimNames);
+    }
+
+    public class DotNetMeter : IMeter
+    {
+        private string libname;
+        private string libver;
+
+        public DotNetMeter(string libname, string libver)
+        {
+            this.libname = libname;
+            this.libver = libver;
+        }
+
+        public Counter CreateCounter(string name)
+        {
+            return new Counter(libname, libver, name);
+        }
+        
+        public Counter CreateCounter(string name, Dictionary<string,string> labels)
+        {
+            return new Counter(libname, libver, name, labels);
+        }
+
+        public Counter CreateCounter(string name, Dictionary<string,string> labels, string[] dimNames)
+        {
+            return new Counter(libname, libver, name, labels, dimNames);
+        }
+
+        public Counter1D CreateCounter(string name, string d1)
+        {
+            return new Counter1D(libname, libver, name, d1);
+        }
+
+        public Counter1D CreateCounter(string name, Dictionary<string,string> labels, string d1)
+        {
+            return new Counter1D(libname, libver, name, labels, d1);
+        }
+
+        public Counter2D<T> CreateCounter<T>(string name, string d1, string d2)
+        {
+            return new Counter2D<T>(libname, libver, name, d1, d2);
+        }
+
+        public Counter2D<T> CreateCounter<T>(string name, Dictionary<string,string> labels, string d1, string d2)
+        {
+            return new Counter2D<T>(libname, libver, name, labels, d1, d2);
+        }
+
+        public Counter3D CreateCounter(string name, string d1, string d2, string d3)
+        {
+            return new Counter3D(libname, libver, name, d1, d2, d3);
+        }
+
+        public Counter3D CreateCounter(string name, Dictionary<string,string> labels, string d1, string d2, string d3)
+        {
+            return new Counter3D(libname, libver, name, labels, d1, d2, d3);
+        }
+
+        public Gauge CreateGauge(string name)
+        {
+            return new Gauge(libname, libver, name);
+        }
+
+        public Gauge CreateGauge(string name, Dictionary<string,string> labels)
+        {
+            return new Gauge(libname, libver, name, labels);
+        }
+
+        public Gauge CreateGauge(string name, params string[] dimNames)
+        {
+            return new Gauge(libname, libver, name, dimNames);
+        }
+
+        public Gauge CreateGauge(string name, Dictionary<string,string> labels, params string[] dimNames)
+        {
+            return new Gauge(libname, libver, name, labels, dimNames);
+        }
+    }
+
+    public class Counter1D

```

Comment:
These weren't in the spec, were you proposing they needed to be created?

Reply:
I was curious about this, too. Was just trying to figure out if this was added by Noah or Victor (as this is a big PR). 🤣

Reply:
Hey, before this, quick question - should I comment on this giant PR?

Reply:
I added these as wrapper to control the prototype that is exposed.  If we did not have this, it is up to the user to match the # of dimension values with the same # when they created it.

Also, a larger item we need to discuss is the flexibility and freedom of adding arbitrary # of labels.  The .NET API currently requires us to declare the # of dim names upfront.  This maybe conflicting with current metric spec.

Reply:
many existing systems (applicationinsights, prometheus-dot-net, msft internal metric system) require dimension keys to be given in advance. Otel doesn't require that -you can add any key/value at any time.

Reply:
@victlu Does knowing ahead the names of dimension keys allow a perf optimization?

Reply:
> The .NET API currently requires us to declare the # of dim names upfront. This maybe conflicting with current metric spec.

I did this as an experiment because I saw Prometheus had done it and the API seemed nice to me. It certainly conflicts with the current spec : )

> @victlu Does knowing ahead the names of dimension keys allow a perf optimization?

Yeah several:
 - don't need to check the key names against the key names in any other label set when determining equality within a particular meter's time series
 - don't need to sort because there is a canonical ordering of the dimension names, 
 - don't need to store the key names as part of a time series label set (lower memory use)

Reply:
Thanks for confirming. I guess these advantages are the reason which motivated all the existing libraries to require label keys in advance.

Reply:
From Otel Metric community meeting, the main reason for *not* requiring labelkeys in advance was to avoid requiring users to always provide the label values,, even when label value is not available.
For example:
```var ctr  = new Counter("countername", "label1","label2");
ctr.Record(100, "label1Value"); // this will throw as label2 is missing.
```

Additionally, one of the goal is to have Baggage or Resource becoming additional labels.
Baggage.Current.Add("user","cijo")
ctr.Record(100, "label1Value", "label2Value"); // this will internally can pick additional labels from the Baggage. (assume SDK is configured to do so).

Reply:
I'm actually challenging pre-defined dim names.  I think we should take a collection of labelset.  A labelset is a collection of key/value pairs.  This requires "merging" of multiple lableset into an effective labelset.  But allows immutability for labelset that is provided (i.e. baggage and attributes, bound instruments, etc...) to be cached.

Reply:
> From Otel Metric community meeting, the main reason for not requiring labelkeys in advance was to avoid requiring users to always provide the label values,, even when label value is not available.

How about?
````
ctr.Record(100, "label1Value", null)
````

Fwiw I'm still not convinced predefined names are the way to go either, but slotting in null doesn't seem that bad? Alternately if the labels at the end of the list were missing that doesn't have to throw an error, the API could just assume that anything missing at the end is automatically null.

> Additionally, one of the goal is to have Baggage or Resource becoming additional labels.

I don't think this conflicts. The label names define the names of labels that the instrumentation code is providing. There is no problem with the SDK adding additional name/value pairs.

> I'm actually challenging pre-defined dim names. I think we should take a collection of labelset.

There is a static labels property which is intended to hold labels where both the name and value are known when the counter is declared. The list of label names are for labels where the value is unknown until the point in time where the measurement is made, it would be impossible to include such a label into a labelset provided to the constructor.